### DISCUSSION TITLE
[Java] What annotations should I use to improve the processing of this predicate

### DISCUSSION THREAD

**Discussion by chmodxxx** [github/codeql#16975] [Q&A]
Hello, 

I have this predicate : 

```
predicate callsFooApi(Method m) {
    exists(delegatedMethod dm |  
     dm.getSignature() = m.getSignature() and 
      (  
         callsOne(dm) or 
         callsTwo(dm) or 
        callsThree(dm)
     )
   )

}

```

So I want to know how I can : 

- short circuit the logic so that if a single delegated Method passes one of the booleans (callsOne .. callsThree), we should stop looking at other delagated Methods. 
- short circuit the actual calls logic so that if callsOne passes, we don't need to verify callsTwo and callsThree 

I know there are a bunch of annotations that codeQL supports, but im not sure how to pick the right ones. 

For reference the delegated method logic is needed since our java code has a bunch of methods overriding each other from difference interfaces and looking at getAnOverride and/or getAnImpl doesn't find all the necessary methods, so resolving methods by signature works in our case.  
Also the callsOne ... callsThree are predicates that perform dataflow and we are ranking them from least resources consuming to most.

**Comment by Yewzir** [comment]


**Reply by chmodxxx** [reply]
I'm specifically looking for compiler pragmas that I can use to force the join ordering direction and not make codeQL arbitrarily join orders
